src/Microsoft.VisualStudio.Composition/ComposedPart.cs

```diff
@@ -160,6 +160,57 @@ public IEnumerable<ComposedPartDiagnostic> Validate(IReadOnlyDictionary<Type, Ex
             }
         }
 
+        /// <summary>
+        /// Checks whether this part contains an import with <see cref="ImportCardinality.ExactlyOne"/> cardinality that has been invalidated,
+        /// which would invalidate this part as well.
+        /// </summary>
+        /// <param name="invalidPartDefnitionsSet">The set of definitions for the invalidated parts.</param>
+        /// <returns>The set of <see cref="ComposedPartDiagnostic"/> with the encountered errors.</returns>
+        public IEnumerable<ComposedPartDiagnostic> CheckForInvalidatedParts(ImmutableHashSet<ComposablePartDefinition> invalidPartDefnitionsSet)
+        {
+            foreach (var pair in this.SatisfyingExports)
+            {
+                var importDefinition = pair.Key.ImportDefinition;
+
+                bool invalidExportFound = false;
+
+                List<ExportDefinitionBinding> newSatisfyingExports = new List<ExportDefinitionBinding>();
+
+                foreach (var export in pair.Value)
+                {
+                    // If the part is invalid it should be removed from the satisfying exports.
+                    if (invalidPartDefnitionsSet.Contains(export.PartDefinition))
+                    {
+                        // Signal that we found an export to remove so that the dictionary is updated
+                        invalidExportFound = true;
+
+                        switch (importDefinition.Cardinality)
+                        {
+                            // Only report error if the cardinality is exactly one
+                            // For mutiple or optional we just remove the satisfying export
+                            case ImportCardinality.ExactlyOne:
+                                yield return new ComposedPartDiagnostic(
+                                    this,
+                                    Strings.RequiredImportHasBeenInvalidated,
+                                    GetDiagnosticLocation(pair.Key),
+                                    GetImportConstraints(pair.Key.ImportDefinition));
+
+                                break;
+                        }
+                    }
+                    else
+                    {
+                        newSatisfyingExports.Add(export);
+                    }
+                }
+
+                if (invalidExportFound)
+                {
+                    this.SatisfyingExports = this.SatisfyingExports.Remove(pair.Key).Add(pair.Key, newSatisfyingExports);
```

Comment:
`Remove` followed by `Add` will create an intermediate `ImmutableDictionary` instance, resulting in several additional allocations. You can make this one operation by using `SetItem`.